### PR TITLE
Create Program

### PR DESCRIPTION
**Pull request by sthitiacharya**

### CODE REVIEW

File: ChallengeTrackerAppAngular/src/app/program/create-new-program/create-new-program.component.ts [hunk 1]
```diff
@@ -68,7 +69,7 @@ export class CreateNewProgramComponent implements OnInit {
 
 		//let startDateString: string = this.parseDate(this.startDate);
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerAppAngular/src/app/program/create-new-program/create-new-program.component.ts:70]: please remove commented code

File: ChallengeTrackerAppAngular/src/app/program/create-new-program/create-new-program.component.ts [hunk 2]
```diff
@@ -68,7 +69,7 @@ export class CreateNewProgramComponent implements OnInit {
...
 		//let targetDateString: string = this.parseDate(this.targetCompletionDate);
-		
+		this.programManager = this.sessionService.getCurrentUser().userId;
 		if (createProgramForm.valid)
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerAppAngular/src/app/program/create-new-program/create-new-program.component.ts:73]: for readability, take out the code inside the if block and return early if createProgramForm is not valid

File: ChallengeTrackerAppAngular/src/app/services/session.service.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  setIsLogin(isLogin: boolean): void {
+    sessionStorage.isLogin = isLogin;
+  }
+
+  getCurrentUser(): User {
+    return JSON.parse(sessionStorage.currentUser);
+  }
+
+  setCurrentUser(currentUser: User | null): void {
+    sessionStorage.currentUser = JSON.stringify(currentUser);
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerAppAngular/src/app/services/session.service.ts:28]: currentUser can be null?

File: ChallengeTrackerAppAngular/src/app/services/user.service.ts [hunk 4]
```diff
@@ -18,6 +18,16 @@ export class UserService {
 
   constructor(private httpClient: HttpClient) { }
 
+  createNewUser(newUser: User): Observable<number> {
+    return this.httpClient.put<number>(this.baseUrl, newUser, httpOptions).pipe(
+      catchError(this.handleError)
+    )
+  }
+  userLogin(username: string | undefined, password: string | undefined): Observable<User> {
+    return this.httpClient.get<User>(this.baseUrl +'/userLogin?username=' + username +'&password=' +
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerAppAngular/src/app/services/user.service.ts:27]: Please don't pass user credentials as query parameters. You can pass them through Authorization header in the http request with basic auth, https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization

@PrasannaVelisetti	Reply…

File: ChallengeTrackerAppAngular/src/app/userManagement/register-user/register-user.component.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    this.resultSuccess = false;
+    this.resultError = false;
+  }
+
+  ngOnInit(): void {
+  }
+
+  create(createUserForm: NgForm){
+    this.submitted = true;
+    if(createUserForm.valid){
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerAppAngular/src/app/userManagement/register-user/register-user.component.ts:35]: for readability, take out the code inside the if block and return early if createUserForm is not valid

File: ChallengeTrackerAppAngular/src/app/userManagement/user-login/user-login.component.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+  userLogin(): void {
+    this.sessionService.setUsername(this.username);
+    this.sessionService.setPassword(this.password);
+
+    this.userService.userLogin(this.username, this.password).subscribe(
+      (response: any) => {
+        let user: User = response;
+
+        if (user != null) {
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerAppAngular/src/app/userManagement/user-login/user-login.component.ts:44]: return early when user is null

File: ChallengeTrackerApplication/ChallengeTrackerApplication-ejb/src/java/ejb/session/stateless/UserEntitySessionBean.java [hunk 7]
```diff
@@ -23,6 +35,41 @@
...
+        if (constraintViolations.isEmpty()) {
+            try {
+
+                em.persist(newUserEntity);
+                em.flush();
+
+                return newUserEntity.getUserId();
+            } catch (PersistenceException ex) {
+                if (ex.getCause() != null && ex.getCause().getClass().getName().equals("org.eclipse.persistence.exceptions.DatabaseException")) {
+                    if (ex.getCause().getCause() != null && ex.getCause().getCause().getClass().getName().equals("java.sql.SQLIntegrityConstraintViolationException")) {
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerApplication/ChallengeTrackerApplication-ejb/src/java/ejb/session/stateless/UserEntitySessionBean.java:60]: so many if statements, please refactor to make intent is clear and for readability

File: ChallengeTrackerApplication/ChallengeTrackerApplication-ejb/src/java/ejb/session/stateless/UserEntitySessionBean.java [hunk 8]
```diff
@@ -47,4 +94,44 @@ public User retrieveUserByUserId(Long userId) throws UserNotFoundException
             throw new UserNotFoundException("User ID " + userId + " does not exist!");
         }
     }
+    
+    @Override
+    public User retrieveUserByUsername(String username) throws UserNotFoundException {
+        Query query = em.createQuery("SELECT u FROM User u WHERE u.username = :inUsername");
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerApplication/ChallengeTrackerApplication-ejb/src/java/ejb/session/stateless/UserEntitySessionBean.java:100]: please move the sql query to a constants file

File: ChallengeTrackerAppAngular/src/app/userManagement/register-user/register-user.component.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+          let newUserId: number = response;
+          this.resultSuccess = true;
+          this.resultError = false;
+          this.message = "New User " + newUserId + " created successfully";
+        },
+        error => {
+          this.resultError = true;
+          this.resultSuccess = false;
+          this.message = "An error has occurred while creating the new user: " + error;
+          console.log('********** CreateNewUserComponent.ts: ' + error);
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerAppAngular/src/app/userManagement/register-user/register-user.component.ts:47]: use string interpolation, for example : `********** CreateNewUserComponent.ts: ${error}`

File: ChallengeTrackerApplication/ChallengeTrackerApplication-war/src/java/ws/rest/MilestoneResource.java [hunk 10]
```diff
@@ -62,7 +62,6 @@ public Response createProgram(CreateMilestoneReq createMilestoneReq)
                 createMilestoneReq.getMilestone().setCreationDate(creationDate);
                 Date date = new SimpleDateFormat("dd-MM-yyyy").parse(createMilestoneReq.getTargetCompletionDate());
                 createMilestoneReq.getMilestone().setTargetCompletionDate(date);
-                createMilestoneReq.getMilestone().setMilestoneCreatedBy(userEntitySessionBean.retrieveUserByUserId(1l));
                 Long milestoneId = milestoneSessionBean.createMilestone(createMilestoneReq.getMilestone(), createMilestoneReq.getProgramId());
                 System.out.println("********** MilestoneResource.createMilestone(): Milestone " + milestoneId + " details passed in via web service");
```
Comment (PrasannaVelisetti) [inline ChallengeTrackerApplication/ChallengeTrackerApplication-war/src/java/ws/rest/MilestoneResource.java:66]: use java string interpolation

File: ChallengeTrackerAppAngular/src/app/services/session.service.ts (hunk 3 above)
Comment (sthitiacharya) [reply ChallengeTrackerAppAngular/src/app/services/session.service.ts:28]: to do

File: ChallengeTrackerAppAngular/src/app/program/create-new-program/create-new-program.component.ts (hunk 1 above)
Comment (sthitiacharya) [reply ChallengeTrackerAppAngular/src/app/program/create-new-program/create-new-program.component.ts]: resolved

File: ChallengeTrackerAppAngular/src/app/program/create-new-program/create-new-program.component.ts (hunk 2 above)
Comment (sthitiacharya) [reply ChallengeTrackerAppAngular/src/app/program/create-new-program/create-new-program.component.ts]: done

File: ChallengeTrackerAppAngular/src/app/services/user.service.ts (hunk 4 above)
Comment (sthitiacharya) [reply ChallengeTrackerAppAngular/src/app/services/user.service.ts]: in progress

File: ChallengeTrackerAppAngular/src/app/userManagement/register-user/register-user.component.ts (hunk 5 above)
Comment (sthitiacharya) [reply ChallengeTrackerAppAngular/src/app/userManagement/register-user/register-user.component.ts]: done

File: ChallengeTrackerAppAngular/src/app/userManagement/user-login/user-login.component.ts (hunk 6 above)
Comment (sthitiacharya) [reply ChallengeTrackerAppAngular/src/app/userManagement/user-login/user-login.component.ts]: done

File: ChallengeTrackerAppAngular/src/app/userManagement/register-user/register-user.component.ts (hunk 9 above)
Comment (sthitiacharya) [reply ChallengeTrackerAppAngular/src/app/userManagement/register-user/register-user.component.ts]: done

File: ChallengeTrackerApplication/ChallengeTrackerApplication-ejb/src/java/ejb/session/stateless/UserEntitySessionBean.java (hunk 7 above)
Comment (sthitiacharya) [reply ChallengeTrackerApplication/ChallengeTrackerApplication-ejb/src/java/ejb/session/stateless/UserEntitySessionBean.java]: done

**Pull request merged**